docs/spec/auth/oauth.md

```diff
@@ -0,0 +1,122 @@
+<!--[metadata]>
++++
+title = "Oauth2 Token Authentication"
+description = "Specifies the Docker Registry v2 authentication"
+keywords = ["registry, on-prem, images, tags, repository, distribution, oauth2, advanced"]
+[menu.main]
+parent="smn_registry_ref"
++++
+<![end-metadata]-->
+
+# Docker Registry v2 authentication using OAuth2
+
+This document describes support for the OAuth2 protocol within the authorization
+server. [RFC6749](https://tools.ietf.org/html/rfc6749) should be used as a
+reference for the protocol and HTTP endpoints described here.
+
+## Refresh token format
+
+The format of the refresh token is completely opaque to the client and should be
+determined by the authorization server. The authorization should ensure the
+token is sufficiently long and is responsible for storing any information about
+long-lived tokens which may be needed for revoking. Any information stored
+inside the token will not be extracted and presented by clients.
+
+## Getting a token
+
+POST /token
+
+#### Headers
+Authorization headers
+Content-Type: application/x-www-form-urlencoded
+
+#### Post parameters
+
+<dl>
+    <dt>
+        <code>grant_type</code>
+    </dt>
+    <dd>
+        (REQUIRED) Type of grant used to get token. When getting a refresh token
+        using credentials this type should be set to "password" and have the
+        accompanying basic auth header. Type "authorization_code" is reserved
+        for future use for authenticating to an authorization server without
+        having to send credentials directly from the client. When requesting an
+        access token with a refresh token this should be set to "refresh_token".
+    </dd>
+    <dt>
+        <code>service</code>
+    </dt>
+    <dd>
+        (REQUIRED) The name of the service which hosts the resource to get
+        access for. Refresh tokens will only be good for getting tokens for
+        this service.
+    </dd>
+    <dt>
+        <code>client</code>
```

Comment:
This was not intending to be the `client_id` from the oauth 2 specification. Oauth2 specifies a client registration process which we do not have. We may want to formalize this though and use `client_id` without an associated secret and just have `docker-engine` or something be a pre-registered value which the hub can add more description around. If this remains free form then it might be allowing the client to set this to any arbitrary value which it will be hard to get any meaning out of.